### PR TITLE
add numpy@1.16.1, patch numpy%intel

### PR DESCRIPTION
**Pull request by svenevs**
- Adds new numpy version 1.16.1
- Patches 1.15.* - 1.16.* releases.  See [`numpy/12991`](https://github.com/numpy/numpy/pull/12991).

I didn't see a good way to do this multiline patch with existing utilities.  Please let me know if there's a cleaner way to do this (see comments, patch file can't be used because the line numbers change).

### CODE REVIEW

**Comment by svenevs** [comment]
Alternatively, I can add a kwargs parameter `re_flags=0` to `filter_file` parameters. Then people could call `filter_file(..., re_flags=(re.I | re.M)` for example. But would I patch that here or would I need to do that in some upstream repo for LLNL and then downstream here?

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 1]
```diff
@@ -173,6 +174,49 @@ def write_empty_libs(f, provider):
                     f.write('libraries=%s\n' % lapackblas_names)
                     write_library_dirs(f, lapackblas_dirs)
 
+        # See: https://github.com/numpy/numpy/pull/12991
+        # Bug introduced: https://github.com/numpy/numpy/pull/10665
+        # So affects 1.15.* => 1.16.* releases
+        # Patch not needed after numpy@1.17.0
+        if '%intel' in spec and spec.satisfies('@1.15.0:1.16.999'):
+            # Need to add #pragma intel optimization_level 2, cannot use
+            # llnl.util.filter_file because this is a multiline match, but a
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Note that in general you should be able to do regular expressions with newlines. `re.MULTILINE` can be used for a more-intuitive handling of `^` and `$`, but I don't see that in your expression - sorry if I missed it. As a sanity check, does the following fail:

```
filter_file(
  r"(NPY_NO_EXPORT void\n"
  r"@TYPE@_@kind@\(char \*\*args, npy_intp \*dimensions, npy_intp \*steps, void \*NPY_UNUSED\(func\)\))",  # noqa: E501
  # Add #pragma patch and paste back function signature
  "#ifdef __INTEL_COMPILER\n"
  "    #pragma intel optimization_level 2\n"
  "#endif\n"
  r"\1",
  filename
)
```

See also:

* https://docs.python.org/2/library/re.html
* https://stackoverflow.com/questions/14689531/how-to-match-a-new-line-character-in-python-raw-string

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 2]
```diff
@@ -26,6 +26,7 @@ class PyNumpy(PythonPackage):
         'numpy.distutils.command', 'numpy.distutils.fcompiler'
     ]
 
+    version('1.16.1', 'dafda51934f645d888866f98424521ae')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Hi @svenevs https://github.com/spack/spack/pull/10738 added 1.16.2 which created a conflict here. Would you be able to address that as well as https://github.com/spack/spack/pull/10647#discussion_r259556762?

**Comment by svenevs** [comment]
The situation is more complicated than anticipated. I'll open another PR once numpy 12991 has a clear direction.

Sorry for the long delay, I appreciate you reviewing this!

**Pull request closed without merging**